src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,163 @@ namespace scene {
+}
+
+SceneNode& SceneNode::createChild(SceneNodeTags childNodeTags) {
+  CORRADE_ASSERT(
+      !(sceneNodeTags_ & SceneNodeTag::Leaf),
+      "SceneNode::createChild(): Can not create child from leaf node", *this);
   // will set the parent to *this
   SceneNode* node = new SceneNode(*this);
   node->setId(this->getId());
+  node->setSceneNodeTags(childNodeTags);
   return *node;
 }
 
+SceneNode& SceneNode::setParent(SceneNode* newParent) {
+  // Perform same internal checks as magnum to ensure newParent is a valid new
+  // parent before updating any SensorSuites
+  // Parent can not be leaf node
+  CORRADE_ASSERT(!(newParent->getSceneNodeTags() & SceneNodeTag::Leaf),
+                 "SceneNode::setParent(): New parent node can not be leaf node",
+                 *this);
+
+  // Skip if this is scene or RootNode (which cannot have parent), or if
+  // newParent is already parent
+  if (isScene() || SceneGraph::isRootNode(*this) ||
+      this->parent() == newParent) {
+    return *this;
+  }
+
+  // Skip if this is an ancestor of newParent
+  auto p = newParent->parent();
+  while (p) {
+    if (p == this)
+      return *this;
+    p = p->parent();
+  }
+
+  // Update old ancestors' SubtreeSensorSuites
+  removeSubtreeSensorsFromAncestors();
+  // Remove sensors from old parent node's nodeSensorSuite
+  removeSensorFromParentNodeSensorSuite();
+
+  MagnumObject::setParent(newParent);
+
+  // Update new ancestors'SubtreeSensorSuites
+  addSubtreeSensorsToAncestors();
+  // Add sensors to newParent's nodeSensorSuite
+  addSensorToParentNodeSensorSuite();
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, add it to parent's nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .add(it->second);
+  }
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, remove it from parent's
+  // nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .remove(it->first);
+  }
+}
+
```

Comment:
Thank you for the comments @bigbike ! Since a leaf node can only have 0 or 1 sensors in its nodeSensorSuite, I changed the logic here a bit to just get the first sensor in the map and add/remove it.

I also call addSensorToParentNodeSensorSuite() in Sensor's constructor now.

Reply:
Can you confirm if this is a valid way to do so? I know that if the map is not empty, getSensors().begin() should be valid, but I still check if it is not end() just in case.

Reply:
Addressed in 1:1 with Yili